src/object.d

```diff
@@ -155,15 +166,271 @@ auto opEquals(Object lhs, Object rhs)
     return lhs.opEquals(rhs) && rhs.opEquals(lhs);
 }
 
-/************************
-* Returns true if lhs and rhs are equal.
-*/
-auto opEquals(const Object lhs, const Object rhs)
-{
-    // A hack for the moment.
+bool opEquals(T, U)(const T lhs, const U rhs)
+    if (is(T == class) && is(U == class) &&
+        __traits(hasMember, T, "opEquals") &&
+        __traits(hasMember, U, "opEquals") &&
+        !is(typeof(lhs.opEquals(rhs)) == bool) &&
```

Comment:
Yes. The whole point of this horrific overload is to cast away `const` when `opEquals` is called on a `const` object to allow for `const` objects to be compared. It's a hack that opens the door for badly behaving `opEquals` implementations breaking `const`, but without it, there's no way to compare `const` objects, because `Object`'s `opEquals` isn't `const`.

With the tempaltized `opEquals`, if the `opEquals` declared on a base class _is_ `const`, then it'll call the main `opEquals` overload without needing this one, and `const` won't be cast away. And when `opEquals` on `Object` is deprecated, we can deprecate this overload. Because `lhs` and `rhs` are `const`, the `!` means that this will only compile if the member function `opEquals` being used isn't `const`. In the long run, once this overload is gone, such a comparison won't even be legal, but we can't do that as long as the `opEquals` on `Object` is still what's being used.